File: projects/advent_of_code/solutions/year2015/day02/go/solution.go
package main

import (
	"fmt"

	"github.com/jomcgi/homelab/projects/advent_of_code/pkg/aoc"
)

func main() {
	input, err := aoc.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

func part1(input string) int {
	total := 0
	for _, line := range aoc.Lines(input) {
		dims := aoc.Ints(line)
		l, w, h := dims[0], dims[1], dims[2]
		sides := []int{l * w, w * h, h * l}
		surfaceArea := 2 * (sides[0] + sides[1] + sides[2])
		extra := min(sides[0], sides[1], sides[2])
		total += surfaceArea + extra
	}
	return total
}

func part2(input string) int {
	total := 0
	for _, line := range aoc.Lines(input) {
		dims := aoc.Ints(line)
		l, w, h := dims[0], dims[1], dims[2]
		perimeters := []int{2*l + 2*w, 2*w + 2*h, 2*h + 2*l}
		ribbon := min(perimeters[0], perimeters[1], perimeters[2])
		bow := l * w * h
		total += ribbon + bow
	}
	return total
}

File: projects/advent_of_code/solutions/year2015/day02/python/solution.py
#!/usr/bin/env python3
"""Advent of Code 2015 Day 2"""

from projects.advent_of_code.python.aoc import lines, ints, read_file


def part1(input_text: str) -> int:
    total = 0
    for line in lines(input_text):
        l, w, h = ints(line)
        sides = [l * w, w * h, h * l]
        surface_area = 2 * sum(sides)
        extra = min(sides)
        total += surface_area + extra
    return total


def part2(input_text: str) -> int:
    total = 0
    for line in lines(input_text):
        l, w, h = ints(line)
        dimensions = sorted([l, w, h])
        ribbon = 2 * dimensions[0] + 2 * dimensions[1]
        bow = l * w * h
        total += ribbon + bow
    return total


if __name__ == "__main__":
    input_text = read_file("../input.txt")

    print(f"Part 1: {part1(input_text)}")
    print(f"Part 2: {part2(input_text)}")
